ident/map.go

```diff
@@ -0,0 +1,256 @@
+	key mapKey
+	// value type stored
+	value ValueType
+}
+
+type mapKey struct {
+	id       ID
+	finalize bool
+}
+
+// Key returns the map entry key.
+func (e MapEntry) Key() ID {
+	return e.key.id
+}
+
+// Value returns the map entry value.
+func (e MapEntry) Value() ValueType {
+	return e.value
+}
+
+// MapOptions is a set of options used when creating an identifier map.
+type MapOptions struct {
+	// Size is the initial size of the map to create.
+	Size int64
+	// Pool is an underlying pool to use when taking copies of the ID
+	// when using the default Set semantics.
+	Pool Pool
+}
+
+// NewMap returns a new identifier map.
+func NewMap(opts MapOptions) *Map {
+	var lookup map[MapHash]MapEntry
+	if opts.Size > 0 {
+		// Only use the constructor with specific size if set to let
+		// go's default size be used if no specific size was set
+		lookup = make(map[MapHash]MapEntry, opts.Size)
+	} else {
+		lookup = make(map[MapHash]MapEntry)
+	}
+	return &Map{
+		lookup: lookup,
+		hash:   hashFnv,
+		opts:   opts,
+	}
+}
+
+func (m *Map) copyKey(key ID) ID {
+	if m.opts.Pool == nil {
+		bytes := append([]byte(nil), key.Data().Get()...)
+		return BinaryID(checked.NewBytes(bytes, nil))
+	}
+	return m.opts.Pool.Clone(key)
+}
+
+// Get returns a value in the map for an identifier if found.
+func (m *Map) Get(id ID) (ValueType, bool) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			return v.value, true
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+	var empty ValueType
+	return empty, false
+}
+
+// Set will set the value for an identifier.
+func (m *Map) Set(id ID, value ValueType) {
+	m.set(id, value, mapSetOptions{
+		copyKey:     true,
+		finalizeKey: true,
+	})
+}
+
+// SetNoCopyNoFinalizeKey will set the value for an identifier but will
+// avoid copying the identifier and will not finalize the key if/when evicted.
+func (m *Map) SetNoCopyNoFinalizeKey(id ID, value ValueType) {
+	m.set(id, value, mapSetOptions{
+		copyKey:     false,
+		finalizeKey: false,
+	})
+}
+
+type mapSetOptions struct {
+	copyKey     bool
+	finalizeKey bool
+}
+
+func (m *Map) set(id ID, value ValueType, opts mapSetOptions) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			m.lookup[hash] = MapEntry{
+				key:   v.key,
+				value: value,
+			}
+			return
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+
+	var keyID ID
+	if opts.copyKey {
+		keyID = m.copyKey(id)
+	} else {
+		keyID = id
+	}
+
+	key := mapKey{
+		id:       keyID,
+		finalize: opts.finalizeKey,
+	}
+
+	m.lookup[hash] = MapEntry{
+		key:   key,
+		value: value,
+	}
+}
+
+// Iter provides the underlying map to allow for using a native Go for loop
+// to iterate the map, however callers should only ever read and not write
+// the map.
+func (m *Map) Iter() map[MapHash]MapEntry {
```

Comment:
Would it be better to call this something like `Iterable`? Feels like a function named `Iter` should return an iterator instead.
```
For k, v := range m.Iterable()
```
seems more intuitive than
```
For k, v := range m.Iter()
```